diff -uNr solfege-3.14.11.old/run-solfege.py.in solfege-3.14.11/run-solfege.py.in
--- solfege-3.14.11.old/run-solfege.py.in	2010-02-08 11:55:13.000000000 +0100
+++ solfege-3.14.11/run-solfege.py.in	2010-03-07 21:13:20.000000000 +0100
@@ -32,6 +32,10 @@
 sys.path.insert(1, os.path.join(prefix, "lib", "solfege"))
 sys.path.insert(1, os.path.join(prefix, "share", "solfege"))
 
+# don't use sys.path[0] which is either the directory containing the script
+# or the empty string which makes python to look into the current directory
+sys.path.pop(0)
+
 import src
 import src.cfg
 from src import filesystem
